yargs.d.ts

```diff
@@ -28,7 +28,7 @@ import { YargsInstance, RebaseFunction } from './build/lib/yargs'
 import { Parser, DetailedArguments, Configuration } from 'yargs-parser/build/lib/yargs-parser-types'
 
 declare namespace yargs {
-  type BuilderCallback<T, R> = ((args: YargsInstance<T>) => PromiseLike<YargsInstance<R>>) | ((args: YargsInstance<T>) => YargsInstance<R>) | ((args: YargsInstance<T>) => void);
+  type BuilderCallback<R1, R2> = ((args: YargsInstance<R1>) => PromiseLike<YargsInstance<R2>>) | ((args: YargsInstance<R1>) => YargsInstance<R2>) | ((args: YargsInstance<R1>) => void);
```

Comment:
what's the motivation for moving away from the variable names for generics defined in `yargs.d.ts`. I was picturing it might be easy to simply pull these over.

Reply:
2 reasons here:
1. T is already used as a generic in several places, so the new T was not usable there (hidden)
2. using T, U, V, etc. for generics makes sense when using them in a few lines long function. However, when you use it application-wide as it is the case here, using letters with more meaning is a necessity to be able to know what you are talking about

I even hesitated using full names, but I finally stuck to one letter generics for readability (it would have lengthen many lines, and people are used to one letter generics).

Reply:
I trust you @mleguen 👍 my concern is  just it becomes a bigger job for you  -- this is outside my comfort zone, with regards to generics.

What I'd like to see, and would be a huge help for my learning, is let's add a TypeScript test suite that literally just exercises type behavior, [like this](https://github.com/yargs/yargs-parser/runs/943541075). Then as the implementation matures, we can write assertions like:

```
it('returns a type of string if .string() is used to configure key')
```

Reply:
@bcoe I would also be much more comfortable if I could rely on a test harness for pointing me out type issues.

However, there are many differences between the "ideal" types described by old @types/yargs now in yargs.d.ts and yargs' code reality, so writing these tests (and having them pass) would not be that straightforward (although necessary, reusing old @types/yargs does is really a pain!).

Sadly, I have to admin I no longer have the necessary bandwidth to be a lead on this...

So let's put yargs.d.ts/yargs.ts aligning in standby for now, until one of us has the time to write such tests.

Reply:
I'm questioning whether it's worth the maintenance burden of pulling in types as complex as `yargs.d.ts` to the library.

It might be better to do without generics when working on the codebase itself, and opt to continue shipping `yargs.d.ts`.

I think this could get us to the goal of a full `ts` conversion, while minimizing risk.

Reply:
I agree with this, but in that case we must keep in mind that:
- `yargs.d.ts` still needs fixing as its signatures does not match exactly the codebase
- every change to `lib/*.ts` would need to be duplicated into `yargs.d.ts` and, without type test harness, we would not have anything broken if we forget to do it (although that would not be worse than when users used `@type/yargs`)
- without test harness, any modification to `yargs.d.ts` would be untested

Reply:
I think we're in agreement. I will start putting some work into building out "typescript" test suite that exercises the types specifically.

I was looking at how DefinitelyTyped approaches this --  it's difficult to use their test runner, but I found this eslint plugin inspired by it:

https://www.npmjs.com/package/eslint-plugin-expect-type

Would it be okay if I close out this PR for now? I would like to put some work into:

1. figuring out ESM/CJS interop.
2. moving towards our next major release.
3. getting back in the swing of addressing a few bug fixes.

I agree with your suggestion of putting this on standby for now while we get this harness in place.